Dashup components and basic dashup widgets

fidsusj opened the pull request:
Please have a look through my developed dashup components and the two sample widgets. 
Buttons can be used either to call an API or to transfer data from one component to another. But I need to work more on the transfer data object and a generic interface for every component. Furthermore sharing stylesheets is not the problem, rather sharing javascript logic like the materialize script is a little bit tricky when trying to only load it once. When having the next version of my concept as well as the appropriate documentation, I will open another pull request. This just serves for you to not have a too complex review in the end and to already use some widgets like the dashup-share-banner to use in the layout mode use case.

raphaelmue commented:
> First of all fix the codacy errors... many of them seem to be very easy to fix like "missing semicolon" or "strings must use doubleqoutes". Normally we would need to add 20 beer to beer list for u now because u caused 20 errors and asked for review. But i think we can limit this to the easy errors so to 6 or 7... what do u think @raphaelmue ?

I'll agree with that 👍

raphaelmue commented on de.dashup.application/src/main/webapp/components/dashup-button.js:
```diff
@@ -0,0 +1,77 @@
+import {LitElement, html, css} from "https://unpkg.com/lit-element@2.1.0/lit-element.js?module";
```
Do you need to import this module in every component, since DashupComponent inherits from LitElement?

fidsusj replied:
Removed the unnecessary imports

raphaelmue commented on de.dashup.application/src/main/webapp/components/dashup-display.js:
```diff
@@ -0,0 +1,29 @@
+import {LitElement, html, css} from "https://unpkg.com/lit-element@2.1.0/lit-element.js?module";
+import {DashupComponent} from "./dashup-component.js";
+export class DashupDisplay extends DashupComponent{
```
Why `DashupDisplay`? A display is something like a monitor or so. I'd change the name to `DashupLabel`

fidsusj replied:
Well, but a label is something that adds descriptive data to other data. The name display is chosen, because the component "displays" data and adds a label to it.

raphaelmue commented on de.dashup.application/src/main/webapp/components/dashup-component.js:
```diff
@@ -0,0 +1,18 @@
+
+    constructor() {
+        super();
+        new CSSStyleSheet().replace("@import url('https://cdnjs.cloudflare.com/ajax/libs/materialize/1.0.0/css/materialize.min.css')").then((sheet) => {
```
You might also bind the files `theme.style.css` and the respective theme, so that within our widgets, the theme is also applied.

fidsusj replied:
Theming and making components resizable is put on the list of tasks for me ;)

raphaelmue commented on de.dashup.application/src/main/webapp/components/dashup-radio-button.js:
```diff
@@ -0,0 +1,19 @@
+import {LitElement, html, css} from "https://unpkg.com/lit-element@2.1.0/lit-element.js?module";
+import {DashupComponent} from "./dashup-component.js";
+export class DashupRadioButton extends DashupComponent{
```
Why does the radio button not have a render function? Is this not implemented yet or not necessary?

fidsusj replied:
Rendering takes place in the Dashup Radio Button Group. These custom elements will be the children of the group, defining how the group should be rendered.

raphaelmue commented on de.dashup.application/src/main/webapp/components/message-broker.js:
```diff
@@ -0,0 +1,13 @@
+class MessageBroker {
```
What is this class for?

fidsusj replied:
This is just a helper class to export logic that does not belong to characteristics of a component to.

raphaelmue commented on de.dashup.application/src/main/webapp/views/includes/head.jsp:
```diff
@@ -28,7 +28,22 @@
+        import '../../components/dashup-list.js'
+        import '../../widgets/dashup-share-banner.js'
+        import '../../widgets/dashup-todo.js'
+    </script>
```
We might consider binding our components only in `index.jsp` and in the future also in the workbench. This improves the performance of all other pages.

fidsusj replied:
I agree.

fidsusj replied:
Fixed

raphaelmue commented on de.dashup.application/src/main/webapp/widgets/dashup-share-banner.js:
```diff
@@ -0,0 +1,45 @@
+import {LitElement, html, css} from 'https://unpkg.com/lit-element@2.1.0/lit-element.js?module'
+export class DashupShareBanner extends LitElement{
```
Is this a widget that a user can add to his dashup? If so, why is this a web component inheriting `LitElement`? I thought we have agreed that our widgets are only plain HTML which contains our dashup components exclusively.

fidsusj replied:
Indeed, this is a widget. You could also leave the surrounding class away, but then you would need to write the html code in the render function inside of your document. This way you would only need to write <dashup-share-banner>. It's just more encapsulated.

fidsusj replied:
See the conversation in issue #85.

raphaelmue commented on de.dashup.application/src/main/webapp/components/dashup-grid-layout.js:
```diff
@@ -0,0 +1,56 @@
+            let offset = element.layout.offset;
+            return html`
+                                <div class="col s${size} m${size} l${size} xl${size} 
+                                            offset-s${offset} offset-m${offset} offset-l${offset} offset-xl${offset}">
```
I think this mechanism is kinda complex when it comes to use `DashupGridLayout`. We should simplify that, for instance leaving the attributes `row` and `offset` out.

fidsusj replied:
A row is definitely a must in order to use a grid layout in another grid layout. This way you can define any custom layout you want. If the user does not need an offset, then he could simply leave it away. I could implement a boolean binding if the offset value has been set.

fidsusj replied:
I made it optionally to set a size and offset for components. The default size is set to 4, whereas default offset is set to 0.

raphaelmue commented on de.dashup.application/src/main/webapp/views/settings.jsp:
```diff
@@ -94,11 +94,11 @@
-                                        <option value="red-love">Red Love</option>
-                                        <option value="white-diamond">White Diamond</option>
-                                        <option value="black-night">Black Night</option>
+                                        <option value="Blue Sky">Blue Sky</option>
```
Why removing the technical name? This technical name is unique and equal to the theme's style sheet.

fidsusj replied:
I did not know why you introduced the technical name until I found out that you bind the stylesheet name to the model, which uses the technical name of course as whitespace is not allowed. Then I rolled it back^^

raphaelmue commented on de.dashup.application/src/test/java/de.dashup.test/steps/LoginRegisterStepdefs.java:
```diff
@@ -106,7 +106,7 @@ public void userCanCloseDashupAndOpenItAgainWithoutBeingLoggedOut() throws IOExc
         HashMap<String, Object> whereParams = new HashMap<>();
         whereParams.put("user_id", "1");
-        Assertions.assertEquals(database.get(Database.Table.USERS_TOKENS, whereParams).getJSONObject(0)
+        Assertions.assertEquals(database.get(Database.Table.TOKENS, whereParams).getJSONObject(0)
```
Why changing the name of the table? Since this table represents the user's tokens, I think this name fits very well.

fidsusj replied:
Can a token refer to anything else but a user? The combination of two table names should only be used for n:m intermediate tables.

raphaelmue commented on de.dashup.model/src/main/java/de/dashup/model/builder/DashupBuilder.java:
```diff
@@ -1,21 +1,29 @@
         StringBuilder content = new StringBuilder();
-        if (user.getSections() != null) {
-            for (Section section : user.getSections()) {
+        List<Section> sections = DashupService.getInstance().loadUserLayout(user).getSections();
```
Another reason why we have introduced the `User` class: We do not want to fetch the data every time again from the database. This is a huge lack of performance.

fidsusj replied:
Again, I did not know, that the user object and the layout is stored locally in the browser. But be aware, that a user can easily manipulate this representation this way. LoadUserLayout generates the data structure for the central dashboard as mentioned in a comment above.

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/DatabaseModels/DatabaseObject.java:
```diff
@@ -0,0 +1,7 @@
+package de.dashup.shared.DatabaseModels;
```
In Java packages are written in lower case.

fidsusj replied:
You are right, my fault.

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/DatabaseModels/DatabaseSectionWidgets.java:
```diff
@@ -0,0 +1,44 @@
+    @SerializedName("widget_id")
+    private Integer widgetID;
+    @SerializedName("predecessor_id")
+    private Integer predecessorID;
```
We decided to remove the `predecessor` and introduce the `order` again. This was also a task of @svenleonhard, which will cause merge conflicts.

fidsusj replied:
As mentioned above, there was a lack of communication, unfortunately.

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/DatabaseModels/DatabaseUser.java:
```diff
@@ -0,0 +1,95 @@
+    private String password;
+    private String salt;
+    @SerializedName("background_image")
+    private String backgroundImage;
```
Why not leaving the settings in the settings? This causes everything that I have implemented in #72 to fail. Furthermore we decided to separate the settings from the user.

fidsusj replied:
See one of the comments above (normalization, implementing performance thoughts, ...)

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/DatabaseModels/DatabaseWidget.java:
```diff
@@ -0,0 +1,91 @@
+    private String overviewText;
+    private String category;
+    @SerializedName("publisher_id")
+    private Integer publisherID;
```
Technically this is a user; we don't have a publisher. Even though a user can publish a widget, a user is not a publisher.

fidsusj replied:
I just thought it was confusing having an attribute "id" and another called "user_id" in the user table when the user_id is the publisher_id.

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/DatabaseModels/DatabaseWidget.java:
```diff
@@ -0,0 +1,91 @@
+    @SerializedName("publisher_id")
+    private Integer publisherID;
+    @SerializedName("publication_date")
+    private Integer publicationDate;
```
Why not using an `int`?

fidsusj replied:
Then you can use it as keys in data structures like the HashMap.

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/DatabaseModels/Order.java:
```diff
@@ -0,0 +1,9 @@
+package de.dashup.shared.DatabaseModels;
+
+public interface Order {
```
Obsolete, since we have decided to delete the predecessor. 
Besides an order describes only how elements are arranged and does not have any of these attributes. And the attribute `id` is already implemented in the `DatabaseObject` interface.

fidsusj replied:
This was just to define an interface for layout objects to be managed by my layout algorithm. So I actually used it for arranging elements. But yes you are right, this is obsolete now.

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/Enums/Size.java:
```diff
@@ -0,0 +1,37 @@
+package de.dashup.shared.Enums;
```
Packages in java...

fidsusj replied:
...are written in lower case.

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/Enums/Size.java:
```diff
@@ -0,0 +1,37 @@
+package de.dashup.shared.Enums;
+
+public enum Size {
```
Why did you remove the size from the widget? I don't remember that something other than a widget can have a size.

fidsusj replied:
The widget class way quite big, I wanted to make the code more modular.

fidsusj replied:
Maybe a section could have a size one day as well.

raphaelmue commented on de.dashup.shared/src/main/java/de/dashup/shared/Enums/Theme.java:
```diff
@@ -0,0 +1,40 @@
+package de.dashup.shared.Enums;
+
+public enum Theme {
```
Same applies here...

fidsusj replied:
As I mentioned in the comment before, I wanted to make the code more modular.

raphaelmue commented on docs/specifications/sad/SAD.md:
```diff
@@ -87,7 +87,7 @@ hands.
 Architectural constraints for this project might be the interaction with widgets, since we do not allow the user to 
 upload executable code for security reasons. So the developer might be limited when it comes to the interaction of the 
-user with the developers widgets. But we do accept this limitation, because of the security goal. 
+user with the developers widgets. But we do accept this limitation, because of the security goal.
```
Why did commit this change? Make sure that you don't commit these changes in the future.

fidsusj replied:
My fault, I didn't recognize that.

raphaelmue commented on docs/specifications/ucs/layout/change_panel_structure/UCS_change_panel_structure.md:
```diff
@@ -27,7 +27,7 @@ dashup - Use Case Specification: Layout - Change Panel Structure
 # 1. Change Panel Structure - Brief Description
 In the use case _change panel structure_ the platform should enable the user to create, modify and delete sections and 
-fill these sections with the desired panels. A rearrangement of sections as well as panels should be possible by drag 
+fill these sections with the desired panelspanels. A rearrangement of sections as well as panels should be possible by drag 
```
Use refactoring carefully. This might happen as well. Please change.

fidsusj replied:
I'm asking myself, what went wrong with the auto refactoring here.

raphaelmue commented on docs/specifications/ucs/marketplace/UCS_marketplace.md:
```diff
@@ -33,7 +33,7 @@ widgets reach from default widgets, that are natively available on the dashup pl
 widget. Furthermore, similar widgets to the current inspected widget, that other users use as well in combination with 
 the current one, are displayed in a <i>similar</i> tab. As a developer it is possible to publish your own widgets to 
-the marketplace from the <i>Workbench</i>. The marketplace is as well the key entry point to add panels to the central 
+the marketplace from the <i>Workbench</i>. The marketplace is as well the key entry point to add panelswidgets to the central 
```
Same applies here...

fidsusj replied:
Agree.

fidsusj commented:
> > First of all fix the codacy errors... many of them seem to be very easy to fix like "missing semicolon" or "strings must use doubleqoutes". Normally we would need to add 20 beer to beer list for u now because u caused 20 errors and asked for review. But i think we can limit this to the easy errors so to 6 or 7... what do u think @raphaelmue ?
> 
> I'll agree with that 👍
> Furthermore you should format all files you have changed. Actually, it would add a few more beers. Please remember this for further pull requests.

I do not want to question IntelliJ, but in some rare cases, the auto format function does not work properly using CRLT+ALT+L. Then I adapt it manually, but normally I take auto formatting into account.

fidsusj commented on de.dashup.application/src/main/java/de/dashup/application/controllers/DashupController.java:
```diff
@@ -12,25 +13,17 @@
-
-            DashupService.getInstance().getSectionsAndPanels(user);
+        return ControllerHelper.defaultMapping(token, request, model, "index", databaseUser -> {
+            DatabaseUser user = DashupService.getInstance().getUserById(databaseUser.getID());
```
My fault, I don't know what came into my mind here.

fidsusj commented on de.dashup.application/src/main/java/de/dashup/application/controllers/EntryController.java:
```diff
@@ -33,16 +33,13 @@ public String handleLogin(@RequestParam(name = "email") String email,
                               @RequestParam(name = "rememberMe", defaultValue = "false") boolean rememberMe,
                               HttpServletRequest request, HttpServletResponse response) throws SQLException {
-        User user = DashupService.getInstance().checkCredentials(email, password, rememberMe);
+        DatabaseUser user = DashupService.getInstance().checkCredentials(email, password, rememberMe);
```
My point was one to remove redudant parsing between the POJO class and the User class and backwards using the fromDatabaseObject method. Two, as long as a user session exists, I wanted to store the user in an instance server side, containing all data. Then I would not need to make a request to the database every time the user requests the server, which would actually increase performance. But if I understood it right, the user object is stored in the local storage of the browser right? Did you consider security issues?

fidsusj commented on de.dashup.application/src/main/java/de/dashup/application/controllers/EntryController.java:
```diff
@@ -33,16 +33,13 @@ public String handleLogin(@RequestParam(name = "email") String email,
         if (user != null) {
             if (rememberMe) {
-                this.localStorage.writeCookie(response, "token", user.getToken());
+                this.localStorage.writeCookie(response, "token", DashupService.getInstance().getTokenByUser(user).getToken());
```
So I guess the user object is generated out of the representation in the cookie? I thought it was genereted using fromDatabaseObject method.

fidsusj commented on de.dashup.application/src/main/java/de/dashup/application/controllers/EntryController.java:
```diff
@@ -33,16 +33,13 @@ public String handleLogin(@RequestParam(name = "email") String email,
-                this.localStorage.writeCookie(response, "org.springframework.web.servlet.i18n.CookieLocaleResolver.LOCALE",
-                        user.getSettings().getLanguage().toLanguageTag());
-            }
+            this.localStorage.writeCookie(response, "org.springframework.web.servlet.i18n.CookieLocaleResolver.LOCALE", user.getLanguage());
```
After having normalized the database scheme, I have merged the settings table into the user table. This saves us additional database requests as our database class does not support inner joins yet and it makes the database scheme more straightforward.

fidsusj commented on de.dashup.application/src/main/java/de/dashup/application/controllers/LayoutModeController.java:
```diff
@@ -54,39 +47,15 @@ public String confirm(@RequestBody DashupPanelStructure body, Model model, HttpS
-                }
-            }
-            lastdss = dss;
+        DatabaseUser user = (DatabaseUser) this.localStorage.readObjectFromSession(request, "user");
```
In order to display the widgets in the dashboard, I needed to generate basic logic fetching the data and generating a data structure for the central dashboard, which wasn't there yet. I changed all the dependencies of Sven's area, but as Sven decided the change the representation of the order of sections and widgets in the database, this would not be applicable anymore.

fidsusj commented on de.dashup.application/src/main/webapp/components/dashup-chart.js:
```diff
@@ -0,0 +1,181 @@
+            <style>
+               ${this.data.map((element, index) => {
+            return html`#q-graph #${element.category} {left: ${index * (80/this.data.length)}%;}`
+        })}
```
This is actually a bug in IntelliJ, isn't it? You do not have to add a comma after each css rule do you. Like rule1{....}, rule2{...}?

joshuaschu replied:
No you do not have to. But I think this is a question of clean code and style and so we should add a comma here. Where is the problem with adding a comma?

fidsusj replied:
I have no problem with adding a comma. I was just wondering why IntelliJ gives such a warning. We don't use commas after any rule in our style sheets of dashup and there it does not give such a warning.

fidsusj commented on de.dashup.application/src/main/webapp/components/dashup-chart.js:
```diff
@@ -0,0 +1,181 @@
+        `;
+    }
+
+    static get styles() {
```
Because the logic using static get styles gets loaded at runtime using the LitElement base class.

fidsusj commented on de.dashup.application/src/main/webapp/components/dashup-grid-layout.js:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    //Called after elements templated is created -> layout for grid layout within a grid layout is already defined
+    firstUpdated() {
```
All of the warnings are caused by the Polymer Framework and the imported LitElement library. The syntax used in the tagged template literals will be parsed to native HTML5 by the tagged html function. IntelliJ does not take this into account. Furthermore, the functions like render are unresolved because they will be added at runtime by the LitElement base class import. (If you try it out in WebStorm after importing LitElement from npm, you should not see a single warning ;) (let me know if it does))

fidsusj replied:
Are you sure I should suppress them? This would add a bunch of // noinspection comments, which in my opinion is not clean code.

joshuaschu replied:
There are annotations like @suppresswarning so you can use that to suppress a single and specified warning. Use that in combination with an explaining comment and everything is fine in my opinion.

fidsusj replied:
As far as I know, the SuppressWarning annotation can only be used in java files. I would need to use //noinspection comment instead in JavaScript files, which would not be compliant to our coding conventions of removing all comments.

fidsusj commented on de.dashup.application/src/main/java/de/dashup/application/controllers/util/ControllerHelper.java:
```diff
@@ -11,12 +13,14 @@
+        if (databaseUser != null) {
+            DatabaseUser user = DashupService.getInstance().getUserById(databaseUser.getID());
+            model.addAttribute("language", user.getLanguage());
+            model.addAttribute("theme", Theme.getThemeByName(user.getTheme()).getTechnicalName());
```
Yes, this is indeed an error, thanks for the report. But the user has been fetched from the local storage anyway. As I was unaware of this I mistakenly requested the user once more from the database. I will fix that.

fidsusj commented on de.dashup.application/src/main/webapp/widgets/dashup-todo.js:
```diff
@@ -0,0 +1,40 @@
+import {LitElement, html} from "https://unpkg.com/lit-element@2.1.0/lit-element.js?module";
+export class DashupTodo extends LitElement{
+
+    render() {
```
See the comments above.

fidsusj commented on de.dashup.model/src/main/java/de/dashup/model/builder/DashupBuilder.java:
```diff
@@ -24,47 +32,45 @@ public static String buildUsersPanels(User user) {
-                    for (Panel p : s.getPanels()) {
+                        .append(section.getID()).append("\">");
+                List<Widget> widgets = section.getWidgets();
+                if (widgets.size() >= 0) {
```
Yes, I need to remove the equals sign here. Thanks!

fidsusj commented on de.dashup.model/src/main/java/de/dashup/model/db/Database.java:
```diff
@@ -170,9 +170,29 @@ public static Database getInstance() throws SQLException {
+                        (JsonDeserializer)(JsonElement json, Type typeOfT, JsonDeserializationContext context) -> {
+                            try{
+                                String idString = json.getAsString();
+                                return "null".equals(idString) ? null : new Integer(idString);
```
I see, I am an old blood Java developer ^^

fidsusj commented:
> Here is my review....
> 
> Checked out your branch in intellij and inspected the code for warnings... found multiple added comments in the code but stoped cause there are so many and this is your job not mine. You should go with the coding conventions and correct these warnings before requesting review. Please correct them or add an explaining comment why you cant fix.
> 
> Executed mvn clean install... unit tests ok. Cucumber tests fail. We cant merge if there are test failures.
> 
> Started application. Layout settings broken. Black sky is the current layout of dashup but layout tab shows blue sky. When i try to change layout I get an 500 error. Does that have anything to do with database scheme? If yes can u please provide the new db scheme as sql statement so we can test the application on your local db? How should we approve a pr without even being able to test the application? I have not checked all features, e.g. registering or stuff but these thinks should work.
> It seems that your "refactoring" broke up many things....
> 
> Formatter is missing for many file checked like three and stoped cause this is your job.
> 
> You have done many changes to the unit tests. The run is ok but my feeling is that the tests are kind of weaker than before... but I cant proof that without a working build and a coverage report....
> 
> This pull request is about widgets and new components. We talked about a example case where we can see these two new panels and the new components in a real dashup in a real widget. I could not find anything like that. Please provide a test user with the panels activated and added. It is not our job as reviewers to build these examples but yours as implementer.
> 
> All in all I totally agree with @raphaelmue. This pr contains many issues (codacy, intellij, even errors in the application, test failures....) so we can not merge this. In addition there are so many changes that are not related to that topic, i see the same issues as Raphael... no idea how to handle that...

All of the warnings are caused by the Polymer Framework and the imported LitElement library. The syntax used in the tagged template literals will be parsed to native HTML5 by the tagged html function. IntelliJ does not take this into account. Furthermore, the functions like render are unresolved because they will be added at runtime by the LitElement base class import. (If you try it out in WebStorm after importing LitElement from npm, you should not see a single warning ;) (let me know if it does))
Are you sure I should suppress them? This would add a bunch of // noinspection comments, which in my opinion is not clean code.
I know I left one comment inside the dashup-button class, I remove it, as you seem to understand the meaning anyway.

I know there are just some tests failing, this pull requests was just meant to gather some general feedback from you, before I continue developing. I thought this would not be a bad idea, before having a lot of things to be reviewed at once.

I will have a look at the change layout settings issue, thanks for the report! Implementing the change layout structure is Svens tasks and is not implemented on this branch so far. I will send you the db scheme via e-mail, sure!

As I said before, I was just my intention to get a review of the things I did so far. Of course, we cannot merge it that way! Sorry for the misunderstanding. But as you see, we already found some mistakes, which would maybe not have been found, if I offered pull requests on a far larger scale!

I will run the formatter on all of the files, sure! I furthermore did not remove any test coverage for existing code. If the code was redundant after my changes, I removed or rewrote the unit tests.

Regarding the test user, please think of my intention behind the pull requests. Once more sorry for the misunderstanding. If you would like to see an example, please have a look at #86 .

All in all, I agree that this pull request contains some refactoring, where I should have created new issues and branches for. I see my mentality of getting stuff done quickly, does not always fit. 
I have an idea of how to handle this pull requests. Let me split it up in smaller pieces first like in #86. Furthermore, I will have a talk with Sven, how far he has come with the layout structure issue. I will probably wait for his commits to be merged in the master, which means I will use the old way of displaying widgets (By creating an HTML file on the server for each widget and size, even if this is not the long term solution (see #88 )). Regarding the performance optimization, we will need to talk about this in our next meeting. But as I said, I urgently needed some feedback, as I did not know if you contended with my changes. Thanks a lot for your in-depth review! Great work!

joshuaschu commented:
I think we can close this pr now or am I wrong?

fidsusj commented:
> To sum it up:
> 
> * There are **again** new issues on codacy. Why don't you fix them before asking us for a review?
> * Changes like refactoring names in our docs, do not belong here.
> * Using always the POJO classes like `DatabaseUser` is a huge lack of performance. If we always try to fetch every additional information from the database, the user waits a long time for the page to reload.
> * Removing the settings and merging its properties into the user's table, is against our decision to separate the settings from the user. Although this is an 1:1 relationship, it is semantically correct.
> * Stick to the naming conventions that are there for Java in general.
> * Inner classes or enums are no violation of coding conventions. Instead it gives the respective class a higher meaning.
> * When changing the dashup_dev database, the other contributors cannot work on this database anymore. Only change this, if a pull request is merged into the master.
> * Just try at least to have successful checks. If you need to change the database scheme, this is not possible, bug changing the code in a way, that even the step "Preparing tests" fails, must not be.
> * And the most important thing: What does all this kind of 'refactoring' have to do with this pull request? The title says that this pull request is for **basic widgets and components** and all that stuff has nothing to do with the title. This pull request actually aimed to create such components and widget, so that all other contributors can continue their work on e.g. the marketplace, layout mode etc. Since this pull request is literally exploded, the other contributors cannot continue their work.
> 
> All in all, I actually cannot approve this pull request. Even I would accept everything but the last one. If we would merge this pull request, there will be tons of merge conflicts and other stuff, like Jenkins failures and so on. I actually do not know how to handle this pr. You can delete every change but the components and widgets. This would be a proper solution. But I actually don't know ...

 First of all, thanks for the extensive feedback!
As I said before, I was just my intention to get a review of the things I did so far. Of course, we cannot merge it that way! Sorry again for the misunderstanding. But I urgently needed some feedback, as I did not know if you contended with my changes. Thanks a lot for your in-depth review! Great work!

- I removed the codacy issues as far as I could, but there are some issues, that I just cannot fix.
- I will take notes, where I have to adapt the docs and do it on a separated branch next time.
- As I mentioned in the comments above, I did not know, that we fetch user details from a browser cookie as an identification method. I will ask you on some security concerns regarding this issue during our next meeting.
- After having normalized the database scheme, I have merged the settings table into the user table. This saves us additional database requests as our database class does not support inner joins yet and it makes the database scheme more straightforward. We should maybe think about our coding conventions once more.
- The idea of giving a class a higher meaning by defining inner enums is well considered. But we need to consider some way of making classes more modular.
- I admit, changing the dev´database directly was not the best idea. Next time I will do it locally first and send you the scheme afterwards.
- See my mentions above. My goal was to receive some general feedback first, I was aware that tests are still failing.  The Jenkins built failed because an empty folder structure does not get pushed. This is why the database credentials file could not be inserted into one of these subfolders as it didn't exist.
- The refactoring commits were made, because I needed to implement a basic infrastructure first to implement the original idea, displaying widgets with components. As we agreed, we wanted to save components directly in the database and not in HTML files on the server for scalability reasons regarding the workbench use case. To implement this, major changes in other files were needed. As I thought opening five other issues and pull requests would cost us too much time, I wanted to handle it in one pull request together, which was a mistake as it seems. I really apologize for this. See issue #88 for the future development. Or see #86 for the components being integrated in dashup the old way.

As I already said for Joshua: 
All in all, I agree that this pull request contains some refactoring, where I should have created new issues and branches for. I see my mentality of getting stuff done quickly, does not always fit.
I have an idea of how to handle this pull request. Let me split it up in smaller pieces first like in #86. Furthermore, I will have a talk with Sven, how far he has come with the layout structure issue. I will probably wait for his commits to be merged in the master, which means I will use the old way of displaying widgets (By creating an HTML file on the server for each widget and size, even if this is not the long term solution (see #88 )). Regarding the performance optimization, we will need to talk about this in our next meeting. But as I said, I urgently needed some feedback, as I did not know if you contended with my changes. Thanks a lot for your in-depth review! Great work!

joshuaschu commented:
> I think we can close this pr now or am I wrong?

?

fidsusj commented:
> > I think we can close this pr now or am I wrong?
> 
> ?

If @raphaelmue agrees, we can close this pull request.
